### PR TITLE
Support pushing beacon node metrics to a remote service

### PR DESCRIPTION
**Pull request by regio**
Implementation to support beacon node metrics publication to a remote service.

#4390 
this implementation complements the enhancement #3773.

## Documentation

- [X] I thought about documentation and added the `documentation` label to this PR if updates are required.

## Changelog

- [X] I thought about adding a changelog entry, and added one if I deemed necessary.

### CODE REVIEW

File: data/publisher/src/main/java/tech/pegasys/teku/data/publisher/MetricsDataFactory.java [hunk 1]
```diff
@@ -33,38 +35,87 @@ public MetricsDataFactory(MetricsSystem metricsSystem) {
...
+    final long network_libp2p_bytes_total_transmit;
+    final int network_peers_connected;
+    final long sync_beacon_head_slot;
+    final String clientVersion;
+
+    if (values.containsKey("head_slot")) {
+      sync_beacon_head_slot =
+          ((Double) ((Observation) values.get("head_slot")).getValue()).longValue();
+    } else {
+      sync_beacon_head_slot = 0;
```
Comment (ajsutton) [inline data/publisher/src/main/java/tech/pegasys/teku/data/publisher/MetricsDataFactory.java]: I'm a little unsure about using 0 as a default here.  I wonder if we should be using the non-primitive `Long` for these values and then if they aren't available, specify `null` (and probably configure the JSON serialiser to ignore nulls so they don't wind up in the pushed JSON at all).  Depends a bit on what the server will do with partial data but sending `0` when we mean unknown may cause a lot of unnecessary alerts.

File: data/publisher/src/main/java/tech/pegasys/teku/data/publisher/MetricsDataFactory.java [hunk 2]
```diff
@@ -33,38 +35,87 @@ public MetricsDataFactory(MetricsSystem metricsSystem) {
...
+    }
+    if (values.containsKey("teku_version")) {
+      clientVersion = ((Observation) values.get("teku_version")).getValue().toString();
+    } else {
+      clientVersion = "";
+    }
```
Comment (ajsutton) [inline data/publisher/src/main/java/tech/pegasys/teku/data/publisher/MetricsDataFactory.java]: Pretty sure `teku_version` will never be available from metrics.  You can just get it directly from `VersionProvider.VERSION` though.

File: data/publisher/src/main/java/tech/pegasys/teku/data/publisher/MetricsDataFactory.java [hunk 3]
```diff
@@ -33,38 +35,87 @@ public MetricsDataFactory(MetricsSystem metricsSystem) {
...
+    if (values.containsKey("peer_count")) {
+      network_peers_connected =
+          ((Double) ((Observation) values.get("peer_count")).getValue()).intValue();
+    } else {
+      network_peers_connected = 0;
+    }
+    return new BeaconNodeMetricData(
+        PROTOCOL_VERSION,
+        System.currentTimeMillis(),
```
Comment (ajsutton) [inline data/publisher/src/main/java/tech/pegasys/teku/data/publisher/MetricsDataFactory.java]: Would be good to inject a `TimeProvider` instance and use that to get the current time rather than calling `System.currentTimeMillis()`.  You can then pass in a `StubTimeProvider` in tests and actually check it's using the right time.  Mostly just makes it consistent with other parts of the system though.

File: data/publisher/src/main/java/tech/pegasys/teku/data/publisher/MetricsDataFactory.java [hunk 4]
```diff
@@ -33,38 +35,87 @@ public MetricsDataFactory(MetricsSystem metricsSystem) {
...
 
-  private static BaseMetricData extractValidatorData(
-      PrometheusMetricsSystem prometheusMetricsSystem) {
+  private static BaseMetricData extractBeaconNodeData(Map<String, Object> values) {
+    final long disk_beaconchain_bytes_total;
+    final long network_libp2p_bytes_total_receive;
+    final long network_libp2p_bytes_total_transmit;
+    final int network_peers_connected;
+    final long sync_beacon_head_slot;
+    final String clientVersion;
```
Comment (ajsutton) [inline data/publisher/src/main/java/tech/pegasys/teku/data/publisher/MetricsDataFactory.java]: We should consistently use camel case for these variables.  Only the fields used to define the JSON API would use snake_case.

File: data/publisher/src/main/java/tech/pegasys/teku/data/publisher/MetricsDataFactory.java [hunk 5]
```diff
@@ -33,38 +35,87 @@ public MetricsDataFactory(MetricsSystem metricsSystem) {
...
+        System.currentTimeMillis(),
+        MetricsDataClient.BEACON_NODE.getDataClient(),
+        disk_beaconchain_bytes_total,
+        network_libp2p_bytes_total_receive,
+        network_libp2p_bytes_total_transmit,
+        network_peers_connected,
+        sync_beacon_head_slot,
+        MetricsDataFactory.CLIENT_NAME,
+        clientVersion,
+        MetricsDataFactory.CLIENT_BUILD);
```
Comment (ajsutton) [inline data/publisher/src/main/java/tech/pegasys/teku/data/publisher/MetricsDataFactory.java]: I suspect `CLIENT_NAME` should just be `VersionProvider.CLIENT_IDENTITY`.  Not sure what `CLIENT_BUILD` is meant to be but suspect it should also be something from `VersionProvider` instead of hard coded here.

File: data/publisher/src/main/java/tech/pegasys/teku/data/publisher/MetricsPublisherManager.java [hunk 6]
```diff
@@ -35,7 +36,7 @@
   private final AsyncRunnerFactory asyncRunnerFactory;
   private final MetricsEndpoint metricsConfig;
   private final MetricsDataFactory dataFactory;
-  private final JsonProvider jsonProvider = new JsonProvider();
+  private final ObjectMapper mapper;
```
Comment (ajsutton) [inline data/publisher/src/main/java/tech/pegasys/teku/data/publisher/MetricsPublisherManager.java]: Why did we stop going via `JsonProvider`?  It would be better if we can continue using it for consistency and to provide some abstraction over the underlying library.

File: data/publisher/src/test/java/tech/pegasys/teku/data/publisher/MetricsDataFactoryTest.java [hunk 7]
```diff
@@ -18,51 +18,50 @@
...
-    assertThat(validatorMetricData.validator_active).isEqualTo(0);
+    final List<BaseMetricData> baseMetricData = metricsDataFactory.getMetricData();
+
+    ObjectMapper mapper = new ObjectMapper();
+    final String data = mapper.writeValueAsString(baseMetricData);
+
+    assertThat(data).contains("beaconnode");
+    assertThat(data).contains("validator");
+    assertThat(data).contains("\"validator_active\":0");
+    assertThat(data).contains("\"disk_beaconchain_bytes_total\":0");
```
Comment (ajsutton) [inline data/publisher/src/test/java/tech/pegasys/teku/data/publisher/MetricsDataFactoryTest.java]: We should deserialise the JSON and then assert on the values rather than trying to do string matches into the JSON (which may fail due to formatting differences).

File: data/publisher/src/test/java/tech/pegasys/teku/data/publisher/MetricsDataFactoryTest.java [hunk 8]
```diff
@@ -18,51 +18,50 @@
...
 import org.hyperledger.besu.metrics.StandardMetricCategory;
 import org.hyperledger.besu.metrics.prometheus.PrometheusMetricsSystem;
 import org.junit.jupiter.api.Test;
 import tech.pegasys.teku.infrastructure.metrics.TekuMetricCategory;
-import tech.pegasys.teku.provider.JsonProvider;
 
 class MetricsDataFactoryTest {
 
-  private JsonProvider jsonProvider = new JsonProvider();
+  //  private final JsonProvider jsonProvider = new JsonProvider();
```
Comment (ajsutton) [inline data/publisher/src/test/java/tech/pegasys/teku/data/publisher/MetricsDataFactoryTest.java]: Shouldn't leave commented out code around but I suspect we should go back to using this.

**Review by ajsutton — COMMENTED** [review]
Generally looks good, just a bunch of tweaks mostly.  Before we consider this story done we'll also need to add tests to ensure the prometheus metrics we're grabbing here actually do exist (ie test should fail if we ever rename any of them).  That will probably need to be an acceptance test that uses a stub external metrics service to publish to and can then check it got non-zero values for everything.

File: data/publisher/src/main/java/tech/pegasys/teku/data/publisher/MetricsDataFactory.java (hunk 2 above)
Comment (regio) [reply data/publisher/src/main/java/tech/pegasys/teku/data/publisher/MetricsDataFactory.java]: Great, I'll use the VersionProvider.IMPLEMENTATION_VERSION instead, because the VERSION property includes also the client name, version, OS and JVM.

File: data/publisher/src/main/java/tech/pegasys/teku/data/publisher/MetricsDataFactory.java (hunk 5 above)
Comment (regio) [reply data/publisher/src/main/java/tech/pegasys/teku/data/publisher/MetricsDataFactory.java]: Thanks a lot for pointing to the VersionProvider class.
Regarding the CLIENT_BUILD, in the specification document, this field is an integer that is "incrementing for easier comparison".  I guess that we don't have this information, right? If so we can omit it.

Here is a snippet of the json example:
` ...  "client_name": "lighthouse",
   "client_version": "1.1.2",
   "client_build": 12, ...`

Comment (ajsutton) [reply data/publisher/src/main/java/tech/pegasys/teku/data/publisher/MetricsDataFactory.java]: Ah right, we don't have that so should just omit it.

**Review by ajsutton — APPROVED** [review]
LGTM.
